Mage.Sets/src/mage/cards/g/GaleWaterdeepProdigy.java

```diff
@@ -0,0 +1,187 @@
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.events.ZoneChangeEvent;
+import mage.game.stack.Spell;
+import mage.players.Player;
+import mage.target.common.TargetCardInYourGraveyard;
+import mage.target.targetpointer.FixedTarget;
+import mage.watchers.common.CastFromHandWatcher;
+
+import java.util.UUID;
+
+/**
+ * @author Rjayz
+ */
+public final class GaleWaterdeepProdigy extends CardImpl {
+
+    public GaleWaterdeepProdigy(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{2}{U}");
+
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.HUMAN);
+        this.subtype.add(SubType.WIZARD);
+        this.power = new MageInt(1);
+        this.toughness = new MageInt(3);
+
+        // Whenever you cast an instant or sorcery spell from your hand,
+        // you may cast up to one of the other type from your graveyard.
+        // If a spell cast from your graveyard this way would be put into your graveyard, exile it instead.
+        this.addAbility(
+                new GaleWaterdeepProdigyTriggeredAbility(new GaleWaterdeepProdigyEffect(),
+                        new FilterInstantOrSorcerySpell("an instant or sorcery spell from your hand"),
+                        false),
+                new CastFromHandWatcher());
+
+        // Choose a Background
+        this.addAbility(ChooseABackgroundAbility.getInstance());
+    }
+
+    private GaleWaterdeepProdigy(final GaleWaterdeepProdigy card) {
+        super(card);
+    }
+
+    @Override
+    public GaleWaterdeepProdigy copy() {
+        return new GaleWaterdeepProdigy(this);
+    }
+}
+
+class GaleWaterdeepProdigyTriggeredAbility extends SpellCastControllerTriggeredAbility {
+
+    public GaleWaterdeepProdigyTriggeredAbility(Effect effect, FilterSpell filter, boolean optional) {
+        super(effect, filter, optional);
+    }
+
+    public GaleWaterdeepProdigyTriggeredAbility(mage.cards.g.GaleWaterdeepProdigyTriggeredAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        if (!super.checkTrigger(event, game)) {
+            return false;
+        }
+
+        CastFromHandWatcher watcher = game.getState().getWatcher(CastFromHandWatcher.class);
+        if (watcher == null || !watcher.spellWasCastFromHand(event.getSourceId())) {
+            return false;
+        }
+
+        Spell spell = game.getState().getStack().getSpell(event.getSourceId());
+        if (spell == null) {
+            return false;
+        }
+
+        FilterCard filterCard = new FilterCard();
+        if (spell.getCardType(game).contains(CardType.SORCERY)) {
+            filterCard.setMessage("an instant card");
+            filterCard.add(CardType.INSTANT.getPredicate());
+        } else {
+            filterCard.setMessage("a sorcery card");
+            filterCard.add(CardType.SORCERY.getPredicate());
+        }
```

Comment:
instead of creating a new filter every time you can just create a static filter for instants and one for sorceries and then choose one of them to use